### PR TITLE
회원 가입 기능 수정

### PR DESCRIPTION
**Pull request by naekang**
## Description

> 회원 가입 기능 수정

## Progress

- [x] 회원 가입 기능 수정
- [x] Member 엔티티 누락된 필드 추가 (name, birth)
- [x] 사용자 정보와 펫 정보 한번에 처리
- [x] 회원 가입 시 입력 정보 Validation
- [x] 테스트 코드

### CODE REVIEW

**Comment by naekang** [comment]
|File|Coverage [79.79%]|:x:|
|:-|:-:|:-:|
|[LocalDateValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/976f25fb701cd8f0664180c2841a1bc39cb3c82e/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FLocalDateValidator.java)|100%|:green_apple:|
|[MemberErrorCode.java](https://github.com/Gaethering/Gaethering-Server/blob/976f25fb701cd8f0664180c2841a1bc39cb3c82e/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fexception%2Ferrorcode%2FMemberErrorCode.java)|100%|:green_apple:|
|[EnumValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/976f25fb701cd8f0664180c2841a1bc39cb3c82e/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FEnumValidator.java)|92.31%|:green_apple:|
|[MemberServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/976f25fb701cd8f0664180c2841a1bc39cb3c82e/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fservice%2FMemberServiceImpl.java)|84.77%|:green_apple:|
|[Member.java](https://github.com/Gaethering/Gaethering-Server/blob/976f25fb701cd8f0664180c2841a1bc39cb3c82e/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fdomain%2FMember.java)|74.77%|:x:|
|[MemberController.java](https://github.com/Gaethering/Gaethering-Server/blob/976f25fb701cd8f0664180c2841a1bc39cb3c82e/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fcontroller%2FMemberController.java)|15.79%|:x:|

|Total Project Coverage|47.07%|:x:|
|:-|:-:|:-:|

**Comment by naekang** [comment]
|File|Coverage [81.75%]|:green_apple:|
|:-|:-:|:-:|
|[LocalDateValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/b2d77432cd487d30d844be7998bac88b5b089b41/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FLocalDateValidator.java)|100%|:green_apple:|
|[MemberErrorCode.java](https://github.com/Gaethering/Gaethering-Server/blob/b2d77432cd487d30d844be7998bac88b5b089b41/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fexception%2Ferrorcode%2FMemberErrorCode.java)|100%|:green_apple:|
|[EnumValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/b2d77432cd487d30d844be7998bac88b5b089b41/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FEnumValidator.java)|92.31%|:green_apple:|
|[MemberServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/b2d77432cd487d30d844be7998bac88b5b089b41/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fservice%2FMemberServiceImpl.java)|84.77%|:green_apple:|
|[Member.java](https://github.com/Gaethering/Gaethering-Server/blob/b2d77432cd487d30d844be7998bac88b5b089b41/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fdomain%2FMember.java)|74.77%|:x:|
|[MemberController.java](https://github.com/Gaethering/Gaethering-Server/blob/b2d77432cd487d30d844be7998bac88b5b089b41/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fcontroller%2FMemberController.java)|44.74%|:x:|

|Total Project Coverage|47.41%|:x:|
|:-|:-:|:-:|

File: src/main/java/com/gaethering/gaetheringserver/member/service/MemberServiceImpl.java [hunk 1]
```diff
@@ -48,32 +56,51 @@ public void confirmEmailAuthCode(String code) {
 
     @Override
     @Transactional
-    public String signUp(SignUpRequest signUpRequest) {
+    public SignUpResponse signUp(MultipartFile file, SignUpRequest signUpRequest) {
+
+        String imageUrl = imageUploader.uploadImage(file);
```
[the remark this answers is not in the document]
Comment (naekang) [reply src/main/java/com/gaethering/gaetheringserver/member/service/MemberServiceImpl.java:61]: 회원가입 전 이메일 보내는 과정에서 이메일 중복 여부를 체크하는 로직이 있어서 제외했습니다!

[the remark this answers is not in the document]
Comment (bokyung95) [reply src/main/java/com/gaethering/gaetheringserver/member/service/MemberServiceImpl.java:61]: 넵!

File: src/main/java/com/gaethering/gaetheringserver/member/service/MemberServiceImpl.java [hunk 2]
```diff
@@ -73,7 +77,26 @@ public String signUp(SignUpRequest signUpRequest) {
...
+            .gender(signUpRequest.getPetGender())
+            .breed(signUpRequest.getBreed())
+            .weight(signUpRequest.getWeight())
+            .isNeutered(signUpRequest.isNeutered())
+            .description(signUpRequest.getDescription())
+            .imageUrl(imageUrl)
+            .isRepresentative(true)
+            .build();
+
+        newPet.setMember(newMember);
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/member/service/MemberServiceImpl.java:96]: 양방향 연관관계로 매핑 되어있으니 Member 쪽에 있는 addPet으로 pet을 추가해주는게 좋을 것 같습니다. 위와 같이 하면 Member쪽에서는 연관관계가 안 맺어져있어서 문제가 될 수도 있을 것 같습니다!

File: src/main/java/com/gaethering/gaetheringserver/member/dto/SignUpRequest.java [hunk 3]
```diff
@@ -34,7 +34,23 @@ public class SignUpRequest {
...
+    private String breed;
+
+    private Gender petGender;
+
+    private String description;
+
+    @JsonProperty("isNeutered")
+    private boolean isNeutered;
+
+    @AssertTrue(message = "비밀번호가 일치하지 않습니다.")
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/member/dto/SignUpRequest.java:61]: 이 부분 말씀드리려고 했는데 해주셨네요! 서비스에서 검사하기보다는 여기서 검사하는게 좋은 방법인 것 같습니다😃

File: src/main/java/com/gaethering/gaetheringserver/member/controller/MemberController.java [hunk 4]
```diff
@@ -23,15 +25,15 @@ public class MemberController {
     private final MemberService memberService;
 
     @PostMapping("/sign-up")
-    public ResponseEntity<SignUpResponse> signUp(@RequestBody @Valid SignUpRequest signUpRequest) {
+    public ResponseEntity<SignUpResponse> signUp(
+        @RequestPart("image") MultipartFile multipartFile,
+        @RequestPart("data") @Valid SignUpRequest signUpRequest
+    ) {
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/member/controller/MemberController.java:31]: 이렇게 받으면 MultipartFile은 form으로 따로 받고, SignUpRequest는 json으로 따로 받게 되는건가요??

File: src/main/java/com/gaethering/gaetheringserver/core/validator/LocalDateValidator.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        try {
+            LocalDate.from(LocalDate.parse(value, DateTimeFormatter.ofPattern(this.pattern)));
+        } catch (DateTimeParseException e) {
+            log.error("DateValidator : {}", e);
+            return false;
+        }
+        return true;
+    }
+}
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/core/validator/LocalDateValidator.java:30]: 유효성 검증할 때 이런식으로 커스텀해서 사용할 수 있군요! 하나 배워갑니다!

File: src/main/java/com/gaethering/gaetheringserver/core/validator/EnumValid.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+@Constraint(validatedBy = EnumValidator.class)
+public @interface EnumValid {
+
+    Class<? extends Enum<?>> enumClass();
+
+    String message() default "must be any of enum {enumClass}";
+
+    Class<?>[] groups() default {};
+
+    Class<? extends Payload>[] payload() default {};
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/core/validator/EnumValid.java:29]: 이 페이로드가 하는 역할은 뭔가요??

File: src/test/java/com/gaethering/gaetheringserver/member/dto/SignUpRequestTest.java [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+            .build();
+
+        //when
+        String message = validator.validate(request).stream().findFirst().get().getMessage();
+
+        //then
+        assertEquals("성별은 MALE, FEMALE 둘 중 하나여야 합니다.", message);
+    }
+
+}
```
Comment (younghoondoodoom) [inline src/test/java/com/gaethering/gaetheringserver/member/dto/SignUpRequestTest.java:411]: 유효성 검증 테스트  코드까지 짜주시는거 너무 좋네요!

File: src/test/java/com/gaethering/gaetheringserver/member/service/MemberServiceTest.java [hunk 8]
```diff
@@ -111,13 +110,13 @@ private static SignUpRequest getSignUpRequest() {
...
             .birth("2017-03-15")
-            .gender(Gender.MALE)
+            .gender("MALE")
             .isEmailAuth(true)
             .petName("뽀삐")
             .petBirth("2022-03-15")
             .weight(5.5f)
             .breed("말티즈")
-            .petGender(Gender.FEMALE)
+            .petGender("FEMALE")
```
Comment (younghoondoodoom) [inline src/test/java/com/gaethering/gaetheringserver/member/service/MemberServiceTest.java:119]: Gender enumtype 없애고 그냥 String으로 가기로 하신건가요?

**Review by younghoondoodoom — COMMENTED** [review]
고생하셨습니다! 테스트 코드까지 꼼꼼하게 짜주셨네요. 궁금한 거랑 제 생각 남겼으니 참고 부탁드립니다!

File: src/main/java/com/gaethering/gaetheringserver/member/dto/SignUpRequest.java (hunk 3 above)
Comment (naekang) [reply src/main/java/com/gaethering/gaetheringserver/member/dto/SignUpRequest.java:61]: ㅎㅎ 좋은 방법 알려주셔서 감사합니다!

File: src/test/java/com/gaethering/gaetheringserver/member/service/MemberServiceTest.java (hunk 8 above)
Comment (naekang) [reply src/test/java/com/gaethering/gaetheringserver/member/service/MemberServiceTest.java:119]: 요청으로 넘어올때는 `String`으로 받고 유효성 검증 후에 `EnumType`으로 변환할 생각입니다!

File: src/main/java/com/gaethering/gaetheringserver/core/validator/EnumValid.java (hunk 6 above)
Comment (naekang) [reply src/main/java/com/gaethering/gaetheringserver/core/validator/EnumValid.java:29]: 사용자 추가 정보를 전달할 수 있는 값으로 주로 심각도를 나타낼때 사용한다고 합니다!
정확히 저희가 어떤 경우에 사용할 수 있는지는 잘 모르겠는데 현재 Validation Annotaion들이 기본적으로 갖고 있는 값이라 추가해뒀습니다!

**Comment by naekang** [comment]
|File|Coverage [73.2%]|:x:|
|:-|:-:|:-:|
|[LocalDateValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/de87bb112fdd5fdbf242b3694b87a7cc1a81da21/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FLocalDateValidator.java)|100%|:green_apple:|
|[MemberErrorCode.java](https://github.com/Gaethering/Gaethering-Server/blob/de87bb112fdd5fdbf242b3694b87a7cc1a81da21/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fexception%2Ferrorcode%2FMemberErrorCode.java)|100%|:green_apple:|
|[EnumValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/de87bb112fdd5fdbf242b3694b87a7cc1a81da21/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FEnumValidator.java)|92.31%|:green_apple:|
|[MemberServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/de87bb112fdd5fdbf242b3694b87a7cc1a81da21/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fservice%2FMemberServiceImpl.java)|85.98%|:green_apple:|
|[Member.java](https://github.com/Gaethering/Gaethering-Server/blob/de87bb112fdd5fdbf242b3694b87a7cc1a81da21/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fdomain%2FMember.java)|75.22%|:x:|
|[MemberController.java](https://github.com/Gaethering/Gaethering-Server/blob/de87bb112fdd5fdbf242b3694b87a7cc1a81da21/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fcontroller%2FMemberController.java)|69.57%|:x:|
|[Comment.java](https://github.com/Gaethering/Gaethering-Server/blob/de87bb112fdd5fdbf242b3694b87a7cc1a81da21/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fboard%2Fdomain%2FComment.java)|3.45%|:x:|

|Total Project Coverage|52%|:x:|
|:-|:-:|:-:|

**Comment by naekang** [comment]
|File|Coverage [73.35%]|:x:|
|:-|:-:|:-:|
|[LocalDateValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/01ecd884cc83894ac7ed0da8471d28d154440d79/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FLocalDateValidator.java)|100%|:green_apple:|
|[MemberErrorCode.java](https://github.com/Gaethering/Gaethering-Server/blob/01ecd884cc83894ac7ed0da8471d28d154440d79/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fexception%2Ferrorcode%2FMemberErrorCode.java)|100%|:green_apple:|
|[EnumValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/01ecd884cc83894ac7ed0da8471d28d154440d79/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FEnumValidator.java)|92.31%|:green_apple:|
|[MemberServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/01ecd884cc83894ac7ed0da8471d28d154440d79/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fservice%2FMemberServiceImpl.java)|86.31%|:green_apple:|
|[Member.java](https://github.com/Gaethering/Gaethering-Server/blob/01ecd884cc83894ac7ed0da8471d28d154440d79/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fdomain%2FMember.java)|75.22%|:x:|
|[MemberController.java](https://github.com/Gaethering/Gaethering-Server/blob/01ecd884cc83894ac7ed0da8471d28d154440d79/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fcontroller%2FMemberController.java)|69.57%|:x:|
|[Comment.java](https://github.com/Gaethering/Gaethering-Server/blob/01ecd884cc83894ac7ed0da8471d28d154440d79/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fboard%2Fdomain%2FComment.java)|3.45%|:x:|

|Total Project Coverage|52.06%|:x:|
|:-|:-:|:-:|

File: src/main/java/com/gaethering/gaetheringserver/member/service/MemberServiceImpl.java (hunk 2 above)
Comment (naekang) [reply src/main/java/com/gaethering/gaetheringserver/member/service/MemberServiceImpl.java]: 넵 수정완료했습니다!

**Comment by naekang** [comment]
|File|Coverage [66.49%]|:x:|
|:-|:-:|:-:|
|[LocalDateValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/708bfd7877a3adc1d1ac9926f85272ff14121881/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FLocalDateValidator.java)|100%|:green_apple:|
|[MemberErrorCode.java](https://github.com/Gaethering/Gaethering-Server/blob/708bfd7877a3adc1d1ac9926f85272ff14121881/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fexception%2Ferrorcode%2FMemberErrorCode.java)|100%|:green_apple:|
|[EnumValidator.java](https://github.com/Gaethering/Gaethering-Server/blob/708bfd7877a3adc1d1ac9926f85272ff14121881/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fcore%2Fvalidator%2FEnumValidator.java)|92.31%|:green_apple:|
|[Member.java](https://github.com/Gaethering/Gaethering-Server/blob/708bfd7877a3adc1d1ac9926f85272ff14121881/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fdomain%2FMember.java)|75.22%|:x:|
|[MemberServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/708bfd7877a3adc1d1ac9926f85272ff14121881/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fservice%2FMemberServiceImpl.java)|54.42%|:x:|
|[MemberController.java](https://github.com/Gaethering/Gaethering-Server/blob/708bfd7877a3adc1d1ac9926f85272ff14121881/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fmember%2Fcontroller%2FMemberController.java)|51.06%|:x:|
|[Comment.java](https://github.com/Gaethering/Gaethering-Server/blob/708bfd7877a3adc1d1ac9926f85272ff14121881/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fboard%2Fdomain%2FComment.java)|3.45%|:x:|

|Total Project Coverage|47.56%|:x:|
|:-|:-:|:-:|

**Pull request merged**